Sources/TuistCore/Graph/GraphTraverser.swift

```diff
@@ -593,6 +609,35 @@ public class GraphTraverser: GraphTraversing {
         return references
     }
 
+    /// Recursively find platform filters within transitive dependencies
+    /// - Parameters:
+    ///   - rootDependency: dependency whose platform filters we need when depending on `transitiveDependency`
+    ///   - transitiveDependency: target dependency
+    /// - Returns: PlatformFilters to apply to transitive dependency
+    func platformFilters(from rootDependency: GraphDependency, to transitiveDependency: GraphDependency) -> PlatformFilters {
+        var visited: Set<GraphDependency> = []
+
+        func find(from root: GraphDependency, to other: GraphDependency) -> PlatformFilters {
+            guard !visited.contains(root) else { return [] }
+            visited.insert(root)
+            guard let dependencies = graph.dependencies[root] else { return [] }
+
+            if dependencies.contains(other) {
+                return graph.edges[(root, other)]
+            } else {
+                let filters = dependencies.map { node in
+                    find(from: node, to: other)
+                }
+
```

Comment:
do we need to account for any intermediate filters in the path?

![Screenshot 2023-10-25 at 09 05 51](https://github.com/tuist/tuist/assets/11914919/4bc167f1-da7d-4b56-970c-2415cddb9a8f)

When the app links/embeds the transitive dependency (Dynamic Framework A) should it also inherit platform filter iOS ? 

I can think of examples where it may be necessary and others where it gets complicated 🙈

For SDK dependencies for example, some don't support all platforms (e.g. CoreTelephony) and in this case we'd want the it to inherit the platform filter 

![Screenshot 2023-11-13 at 13 49 27](https://github.com/tuist/tuist/assets/11914919/50f4fee4-dfa2-4faf-9886-962d3248f948)

However then we'd need to account for all paths if the app also directly depends on the framework then there's no need to inherit the platform filter

![Screenshot 2023-10-25 at 09 05 59](https://github.com/tuist/tuist/assets/11914919/c6a335cd-c940-465d-a82f-56578f32caae)

Reply:
Here's the test case in case its helpful

```swift
func test_platformFilters_transitiveDependencyInheritedPlatformFilter() throws {
        // Given
        let app = Target.test(name: "App", destinations: [.iPad, .iPhone, .mac], product: .app)
        let staticFramework = Target.test(
            name: "StaticFramework",
            destinations: [.iPad, .iPhone],
            product: .staticLibrary
        )
        
        let project = Project.test(targets: [app, staticFramework])
        
        let appkGraphDependency: GraphDependency = .target(name: app.name, path: project.path)
        let staticFrameworkGraphDependency: GraphDependency = .target(name: staticFramework.name, path: project.path)
        let sdkGraphDependency: GraphDependency = .testSDK(name: "CoreTelephony.framework")
        
        // Given: Value Graph
        let graph = Graph.test(
            path: project.path,
            projects: [project.path: project],
            targets: [
                project.path: [
                    app.name: app,
                    staticFramework.name: staticFramework,
                ]
            ],
            dependencies: [
                appkGraphDependency: [
                    staticFrameworkGraphDependency,
                ],
                staticFrameworkGraphDependency: [
                    sdkGraphDependency,
                ]
            ],
            edges: [
                GraphEdge(from: appkGraphDependency, to: staticFrameworkGraphDependency): [.ios],
            ]
        )
        let subject = GraphTraverser(graph: graph)
        
        // When
        let results = subject.platformFilters(from: appkGraphDependency, to: sdkGraphDependency)
        
        // Then
        XCTAssertEqual(results.sorted(), [
            .ios
        ])
    }
```

Reply:
Should we in scenarios like the one above where we have multiple branches reaching the same transitive dependency, find the filters that overlap and use them? Or would we be assuming too much if we do that?

Reply:
Pushed this test and a change that should address this case.